### PR TITLE
kie-issues#1001: Copy the `kogito-swf-{devmode,builder}` images from kogito-images to `kie-tools`

### LINKED ISSUE #1001: Copy the `kogito-swf-{devmode,builder}` images from `kogito-images` to `kie-tools`.
As the `devmode` image for SWF depends on the Quarkus Dev UIs, and the Quarkus Dev UIs are moving to `kie-tools`, we can't have `kogito-images` depending on `kie-tools`, as the `kn-plugin-workflow` package on `kie-tools` already depends on the `devmode` image.

Create a new package at `packages/sonataflow-images` containing the necessary code to build `kogito-swf-builder` and `kogito-swf-devmode` using the same tech-stack present on https://github.com/apache/incubator-kie-kogito-images. Don't forget to add the necessary build command to `README.md` and the `kieTools` property of package.json.

### PR DESCRIPTION
**Pull request by pefernan**
Closes https://github.com/apache/incubator-kie-issues/issues/1001

### CODE REVIEW

File: packages/kogito-swf-devmode/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+This package contains the `cekit` image descriptor (`kogito-swf-devmode-image.yaml`) and modules needed to build the `kogito-swf-devmode`
+image along with the modules and scripts provided in `@kie-tools/kogito-swf-common`.
+
+## Additional requirements
+
+- **python3** with the following packages installed:
+  - `behave` `lxml` `docker` `docker-squash` `elementPath` `pyyaml` `ruamel.yaml` `python-dateutil` `Jinja2` `pykwalify` `colorlog` `click`
+- **cekit 4.11.0**: [docs.cekit.io](https://docs.cekit.io/en/latest/index.html)
+- **s2i**: [source-to-image](https://github.com/openshift/source-to-image)
```
Comment (ricardozanini) [inline packages/kogito-swf-devmode/README.md:11]: We don't need this.

File: scripts/bootstrap/check_required_preinstalled_cli_commands.mjs [hunk 2]
```diff
@@ -31,8 +31,9 @@ const argsByCommand = new Map([
   ["go", ["version"]],
   ["make", ["-v"]],
   ["helm", ["version"]],
-  ["python", "--version"],
-  ["pip", "--version"],
+  ["python", ["--version"]],
+  ["pip", ["--version"]],
+  ["s2i", ["version"]],
```
Comment (ricardozanini) [inline scripts/bootstrap/check_required_preinstalled_cli_commands.mjs:36]: We can remove this requirement!

**Review by ricardozanini — COMMENTED** [review]
Looks pretty good! Congrats! We can reuse this work later when we move the builder/devmode images back to the operator repo.

Just one minor thing, s2i is not needed for any of these images 👍  

So one less component to take care of.

**Comment by tiagobento** [comment]
I think CEKit called `s2i` during the build... I tried removing it and it didn't work :( 

@ricardozanini 🎉 What parts you're talking about reusing?

**Comment by ricardozanini** [comment]
@tiagobento I don't have s2i installed locally and works fine. 🤔

**Comment by tiagobento** [comment]
@ricardozanini Interesting.. Maybe `s2i` is only necessary for tests? See the error we got:
https://github.com/apache/incubator-kie-tools/actions/runs/8744553065/job/23997698651

```
2024-04-18 21:34:02,469 cli.py:542        ERROR CEKit dependency: 's2i' was not found, please provide the 's2i' executable.
```

**Comment by tiagobento** [comment]
@pefernan The error that's happening on the macOS build is also happening on my local machine. I'm using Rancher Desktop... No luck with Colima too (this is what we use on the CI here)....

```
2024-04-19 15:57:31,235 docker_builder.py:224        ERROR Could not connect to the Docker daemon at 'http+docker://localhost', please make sure the Docker daemon is running.
2024-04-19 15:57:31,235 cli.py:542        ERROR Cannot connect to Docker daemon
```

**Comment by klbowers** [comment]
Thanks for the PR! It might be a good idea to add python (and python venv) to the requirements in the main README, as running pnpm bootstrap without them causes errors.
Findings when trying to build on WSL (Ubuntu 22.04):

**kogito-swf-devmode:**
After following steps in the readme, I ran pnpm -F @kie-tools/kogito-swf-devmode build:prod and got the following errors: 
`Installed Bats to bats/bin/bats
----> running bats
/home/kbowers-ibm/test/incubator-kie-tools/packages/kogito-swf-devmode/build/bats/libexec/bats-core/bats: line 103: cd: modules/kogito-swf/common/scripts/tests: No such file or directory
make: *** [Makefile:80: bats] Error 1
make: Leaving directory '/home/kbowers-ibm/test/incubator-kie-tools/packages/kogito-swf-devmode/build'
 ELIFECYCLE  Command failed with exit code 2.
[run-script-if] Error executing 'run-script-os'.
 ELIFECYCLE  Command failed with exit code 1.
`/home/kbowers-ibm/test/incubator-kie-tools/packages/kogito-swf-devmode:`

**kogito-swf-builder:**
Should the builder and devmode have identical readme's? Judging by the errors below it seems like some additional setup might be required.
After following steps in the readme, I ran pnpm -F @kie-tools/kogito-swf-builder build:prod and got the following errors: 
`2024-04-22 16:15:19,784 steps.py:192        WARNING Exception caught: ConnectionError(MaxRetryError("HTTPConnectionPool(host='172.17.0.2', port=8080): Max retries exceeded with url: /q/health/ready (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f2e351d0250>: Failed to establish a new connection: [Errno 111] Connection refused'))"))`

` [io.qua.config] (main) Unrecognized configuration key "kogito.data-index.health-enabled" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
[docker-java-stream--1709125771] INFO RunTests - STDOUT: 2024-04-22 14:18:53,313 WARN  [io.qua.config] (main) Unrecognized configuration key "kogito.jobs-service.health-enabled" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
[docker-java-stream--1709125771] INFO RunTests - STDOUT: 2024-04-22 14:18:55,030 WARN  [org.kie.kog.add.qua.kna.eve.KnativeEventingConfigSourceFactory] (main) K_SINK variable is empty or doesn't exist. Please make sure that this service is a Knative Source or has a SinkBinding bound to it.
[docker-java-stream--1709125771] INFO RunTests - STDOUT: 2024-04-22 14:18:55,976 WARN  [io.sma.rea.mes.provider] (main) SRMSG00207: Some components are not connected to either downstream consumers or upstream producers:
[docker-java-stream--1709125771] INFO RunTests - STDOUT:        - Emitter{channel:'kogito-job-service-job-request-events'} has no downstream
[docker-java-stream--1709125771] INFO RunTests - STDOUT:
[docker-java-stream--1709125771] INFO RunTests - STDOUT: 2024-04-22 14:18:56,540 INFO  [org.kie.kog.add.qua.mes.com.QuarkusKogitoExtensionInitializer] (main) Registered Kogito CloudEvent extension
[docker-java-stream--1709125771] INFO RunTests - STDOUT: 2024-04-22 14:18:56,651 WARN  [io.sma.rea.mes.provider] (vert.x-eventloop-thread-6) SRMSG00234: Failed to emit a Message to the channel: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:9090
[docker-java-stream--1709125771] INFO RunTests - STDOUT: Caused by: java.net.ConnectException: Connection refused`

**Comment by tiagobento** [comment]
@kbowers-ibm Thanks!! Great point about the README. cc @pefernan. 

As for the build errors, did you run

- `pnpm -F @kie-tools/kogito-swf-devmode build:prod`; or
- ~`pnpm -F @kie-tools/kogito-swf-devmode build:prod...`~
- `pnpm -F @kie-tools/kogito-swf-devmode... build:prod`

?

UPDATE: Duh, placed the `...` in the wrong place.

**Comment by klbowers** [comment]
Just copied the command in the kogito-swf-devmode README which is:
pnpm -F @kie-tools/kogito-swf-devmode build:prod
**Update:**
Just tried running the command with 3 dots now and got the same result:
pnpm -F @kie-tools/kogito-swf-devmode... build:prod

**Comment by pefernan** [comment]
> @pefernan The error that's happening on the macOS build is also happening on my local machine. I'm using Rancher Desktop... No luck with Colima too (this is what we use on the CI here)....
> 
> ```
> 2024-04-19 15:57:31,235 docker_builder.py:224        ERROR Could not connect to the Docker daemon at 'http+docker://localhost', please make sure the Docker daemon is running.
> 2024-04-19 15:57:31,235 cli.py:542        ERROR Cannot connect to Docker daemon
> ```

Today I had this issue in a brand new environment, but after the first failed build it started working. Will be taking a look

**Comment by pefernan** [comment]
> Just copied the command in the kogito-swf-devmode README which is: pnpm -F @kie-tools/kogito-swf-devmode build:prod **Update:** Just tried running the command with 3 dots now and got the same result: pnpm -F @kie-tools/kogito-swf-devmode... build:prod

I think the problem is that I **shamefully** forgot copying the test assets in the devmode image... something we may have in mind is that we may have to do some changes in the tests when upgrade to Quarkus `3.8` because there are some tests related to the dev-ui urls that will need to be revisited.

@kbowers-ibm @tiagobento ^

**Comment by tiagobento** [comment]
Heeey look at that! We have a green build on Ubuntu! Now it's up to us to decide to NOT build this on macOS or fix the issue... Of course I think we should fix the issue, but I'm not sure if this is going to bee too hard or not.. Waiting on your assessment @pefernan and @thiagoelg

**Comment by tiagobento** [comment]
@pefernan Oh, one more thing... One of the two images is using a DEBUG option that is really polluting the logs. Do you mind checking which one of them is and changing is to a default INFO/WARN logging level please?

**Comment by klbowers** [comment]
Thanks Pere, running better on my end now! Still getting the problem with warning connection refused, etc. but think that's more a problem with my docker config rather than anything here, especially as the Ubuntu build is now green. 👍

**Comment by tiagobento** [comment]
Due to time constraints, I'm skipping the build on macOS too... :/

**Review by thiagoelg — APPROVED** [review]

**Review by tiagobento — APPROVED** [review]

**Comment by ricardozanini** [comment]
> @ricardozanini Interesting.. Maybe `s2i` is only necessary for tests? See the error we got: https://github.com/apache/incubator-kie-tools/actions/runs/8744553065/job/23997698651
> 
> ```
> 2024-04-18 21:34:02,469 cli.py:542        ERROR CEKit dependency: 's2i' was not found, please provide the 's2i' executable.
> ```

Even for tests, the images using it were removed.

**Comment by tiagobento** [comment]
@ricardozanini Can you please check if the changes that removed `s2i` are present on the `kie-tools` copy of them? See https://kie.zulipchat.com/#narrow/stream/262892-serverless-workflow/topic/Status.20of.20.60kogito-swf-.7Bbuilder.2Cdevmode.7D.60.20images.20.20.40.20KIE.20Tools

**Comment by ricardozanini** [comment]
@tiagobento the builder and devmode images **don't use s2i**.

**Comment by tiagobento** [comment]
@ricardozanini The build failed because `s2i` was not present. We installed it so we can move forward.

I'm not familiar with the code, neither is @thiagoelg, or @pefernan.

If not having `s2i` installed is a requirement, please take a look at the code and provide us instructions on how to proceed so we can build the stuff without `s2i` being installed.

We'd be happy to assist you if you'd like to send a PR addressing this issue yourself too.

**Pull request merged**